File: app/src/main/java/edu/schoolapp/TodoAdapter.java
package edu.schoolapp;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import edu.schoolapp.Model.TodoModel;

public class TodoAdapter extends BaseAdapter {

    private List<TodoModel> todoList;

    TodoAdapter(List<TodoModel> todoList) {
        this.todoList = todoList;
    }

    @Override
    public int getCount() {
        return todoList == null ? 0 : todoList.size();
    }

    @Override
    public Object getItem(int position) {
        return todoList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if (convertView == null) {
            convertView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_todo, parent, false);
            TodoViewHolder viewHolder = new TodoViewHolder(convertView);
            convertView.setTag(viewHolder);
            viewHolder.setupData((TodoModel) getItem(position));
            return convertView;
        }
        ((TodoViewHolder) convertView.getTag()).setupData((TodoModel) getItem(position));
        return convertView;
    }

    public void updateData(List<TodoModel> list) {
        if (todoList == null) {
            todoList = new ArrayList<>();
        } else {
            todoList.clear();
        }
        todoList.addAll(list);
        notifyDataSetChanged();
    }

    static class TodoViewHolder {
        TextView titleTextView;
        TextView descriptionTextView;
        TextView dateTextView;

        TodoViewHolder(View view) {
            this.titleTextView = view.findViewById(R.id.title_text_view);
            this.descriptionTextView = view.findViewById(R.id.description_text_view);
            this.dateTextView = view.findViewById(R.id.date_text_view);
        }

        void setupData(TodoModel data) {
            //TODO setup data here
            titleTextView.setText(data.getTitle());
            descriptionTextView.setText(data.getDescription());
            dateTextView.setText(data.getDate());
        }
    }
}

File: app/src/main/java/edu/schoolapp/TodoAdapter.kt
package edu.schoolapp

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import edu.schoolapp.model.TodoModel

class TodoAdapter internal constructor(private var todoList: MutableList<TodoModel>?) :
    BaseAdapter() {
    override fun getCount(): Int {
        return if (todoList == null) 0 else todoList!!.size
    }

    override fun getItem(position: Int): Any {
        return todoList!![position]
    }

    override fun getItemId(position: Int): Long {
        return 0
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        var convertView = convertView
        if (convertView == null) {
            convertView =
                LayoutInflater.from(parent?.context).inflate(R.layout.item_todo, parent, false)
            val viewHolder = TodoViewHolder(convertView)
            convertView.tag = viewHolder
            viewHolder.setupData(getItem(position) as TodoModel)
            return convertView
        }
        (convertView.tag as TodoViewHolder).setupData(getItem(position) as TodoModel)
        return convertView
    }

    fun updateData(list: List<TodoModel>?) {
        if (todoList == null) {
            todoList = ArrayList()
        } else {
            todoList!!.clear()
        }
        todoList!!.addAll(list!!)
        notifyDataSetChanged()
    }

    internal class TodoViewHolder(view: View) {
        private var titleTextView: TextView = view.findViewById(R.id.title_text_view)
        private var descriptionTextView: TextView = view.findViewById(R.id.description_text_view)
        private var dateTextView: TextView = view.findViewById(R.id.date_text_view)

        fun setupData(data: TodoModel) {
            //TODO setup data here
            titleTextView.text = data.title
            descriptionTextView.text = data.description
            dateTextView.text = data.date
        }
    }
}